Semkiv/admin update fix

min4uk opened the pull request:
Fixed update error by changing incoming Dto.
Removed institutionId changing, because it`s not allowed by business rules.
Removed some code smell.

min4uk committed:
Removed code smell, removed 'ministry' from log and message because this service updates different admins.

min4uk committed:
Removed changing of InstitutionId, because nobody can`t change instituion of administrators, it is bussines rule

min4uk committed:
remove codesmell, fixed summary

min4uk committed:
Added UpdateAdminBaseDto for area,ministry and region admin update. Added new Dto to update of area admin

min4uk committed:
Replaced UpdateAdminBaseDto with BaseUserDto

min4uk committed:
Added BaseUserDto instead of MinistryAdminDto and RegionAdminDto to update method. Fixed tests

VadymLevkovskyi commented:
You've mentioned "Removed institutionId changing, because it`s not allowed by business rules". Are there any tests enforcing this business rule? If not - could they be written?

min4uk commented:
> You've mentioned "Removed institutionId changing, because it`s not allowed by business rules". Are there any tests enforcing this business rule? If not - could they be written?

I will think about how to check it correctly and do it

min4uk commented:
> You've mentioned "Removed institutionId changing, because it`s not allowed by business rules". Are there any tests enforcing this business rule? If not - could they be written?

I tried to, but I couldn't figure out how to write such unit test😥

VadymLevkovskyi commented:
> > You've mentioned "Removed institutionId changing, because it`s not allowed by business rules". Are there any tests enforcing this business rule? If not - could they be written?
> 
> I tried to, but I couldn't figure out how to write such unit test😥

@min4uk please correct me if I'm wrong. As I can see the logic of Id hiding is hidden in `CreateMap<BaseUserDto, NnnAdminBaseDto>` that is used while creating request via Http using `CommunicationService.SendRequest` method.

As `CommunicationService` is parent class, not a dependency (as it has to be!) and mentioned method cannot be mocked we could go deeper and find that real http request is sent via httpClient that is created by `httpClientFactory` - injected dependency that could be mocked.

So one possible solution - is to play with `httpClientFactory` dependency.
Other solution - redesign `NnnAdminService`s to have `CommunicationService` (or part of it related to SendRequest) as a dependency that could be mocked.

min4uk commented:
> > > You've mentioned "Removed institutionId changing, because it`s not allowed by business rules". Are there any tests enforcing this business rule? If not - could they be written?
> > 
> > 
> > I tried to, but I couldn't figure out how to write such unit test😥
> 
> @min4uk please correct me if I'm wrong. As I can see the logic of Id hiding is hidden in `CreateMap<BaseUserDto, NnnAdminBaseDto>`
@VadymLevkovskyi 
Our controller binds only that properties, that corresponds to the properties of `BaseUserDto`, so `institutionId` property will not be binded
![1](https://github.com/ita-social-projects/OoS-Backend/assets/68245598/6eb581c6-ec6f-4747-aedd-8a3a295d4eab)

![2](https://github.com/ita-social-projects/OoS-Backend/assets/68245598/1548aec4-bcdd-4e98-81cb-0db9b96626f6)
Unless we add this property to `BaseUserDto` class or change Dto that we are using.
We have a test that checks that we are using right class `Update_WithInvalidModel_ReturnsRequestObjectResult()` (it has to be ...BadRequestObjectResult() i will fix it)
It seems that I need to check if `institutionId` property have been added to `BaseUserDto`. Can i do it this way?
![3](https://github.com/ita-social-projects/OoS-Backend/assets/68245598/22b17f22-e1bc-4fcf-8fd0-b21e919db42d)

VadymLevkovskyi commented:
@min4uk re suggested test

Long story short: it's more palliative care than the real treatment.

Details.
- Your test does what it states, and that's cool, but not enough.
Your test is really checking if property `institutionId` is not added to `BaseUserDto`. But that time when it fail you will have no idea why it was important to write this unit test. (There is no err message in neither `Assert` and test name is just doubling the code logic, not the real explanation why test was written)

- "Our controller binds only that properties, that corresponds to the properties of `BaseUserDto`" 
What if controller input parameter type will be changed? Suggested `Update_WithInvalidModel_ReturnsRequestObjectResult` test doesn't cover this case.

- So main issue of this test case is that you are checking not the controller action, but (only partially) details of implementation. That's good, but not sufficient.

min4uk committed:
added mapper tests

min4uk committed:
added test that checks that update works only with BaseUserDto

min4uk committed:
added error message

min4uk commented:
@VadymLevkovskyi 
I added error messages to mapper tests, and added `Update_WorkOnlyWithBaseUserDto` test to avoid changing BaseUserDto, without undestanding consequences.
Is this ok, or did you mean that these tests are not enough in any case and i need to moq request sending somehow?

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/RegionAdminControllerTests.cs:
```diff
@@ -210,17 +210,36 @@ public async Task Update_WithInvalidModel_ReturnsRequestObjectResult()
+    public async Task Update_WorkOnlyWithBaseUserDto()
+    {
+        // Arrange
+        var updateRegionAdminDto = new BaseUserDto();
```
Sorry, but I can't agree with this test existence. Probably better delete it.

Details:

The test promises to `_WorkOnlyWithBaseUserDto` but when I change it in test to
`var updateRegionAdminDto = new RegionAdminDto();` 
it doesn't fail. 

If I change it in `UpdateRegionAdminAsync` parameter type, the compiler will error on test code and - see my previous comments - next developer wil just rewrite test to meet the new type.

As for me this checks only one thing - that controller calls service. The same thing that is already verified in 2 tests below.

The initial task - checking absence of "institutionId changing, because it`s not allowed by business rules" is not performed. 

For this check I expect to see in test something like:
```
//Arrange
SeedRepository(new RegionAdminDto() { InstitutionId = oldInstitutionId });
...

//Act
await regionAdminController.Update(new RegionAdminDto() { InstitutionId = otherInstitutionId });

//Assert
var newInstitutionId = GetFromRepository().InstitutionId;
Assert.That(oldInstitutionId, Is.EqualTo(newInstitutionId)); // thus we ensure that it is not changed to otherInstitutionId 
```
Ofc, it's pseudocode as `SeedRepository()` and `GetFromRepository()` are not written and they could be rather big redesign task. It's doable, but I given up to insist on it, and now ask only for this test deletion.

min4uk replied:
@VadymLevkovskyi 
Sorry😓, I want to do this right, not just close PR. Maybe we have misunderstood each other.
Earlier I tried to create such a test as you showed, but in-memory database does not support transactions. We discussed this on meeting, and decided, to try something else. Then i tried to mock send method, but 
> Non-overridable members (here: HttpClient.SendAsync) may not be used in setup / verification expressions.

I can try another option, mock communication service(by changing inheritance with constructor dependency), I don't know any other options

min4uk replied:
Sorry😓, I want to do this right, not just close PR. Maybe we have misunderstood each other.
Earlier I tried to create such a test as you showed, but in-memory database does not support transactions. We discussed this on meeting, and decided, to try something else. Then i tried to mock send method, but

> Non-overridable members (here: HttpClient.SendAsync) may not be used in setup / verification expressions.

**UPD.**
It appeared that i can simply disable that error of in-memory database...😔
Please, check if this test fits

VadymLevkovskyi replied:
Balissimo! Thank you for being persistent.
There is few more percents of test coverage needed to reach the perfection - please check [Sonar report](https://sonarcloud.io/component_measures?id=ita-social-projects-oos-backend&pullRequest=1272&metric=new_coverage&view=list)

min4uk replied:
TY for your patience. I will work on test coverage now

min4uk committed:
deleted test

min4uk committed:
added tests to check that institutionId and CATOTTGId not changed after update

min4uk committed:
added areaAdmin tests

min4uk committed:
Merge branch 'develop' into Semkiv/AdminUpdateFix

min4uk committed:
added negative tests

min4uk committed:
removed getting context to separate method

min4uk committed:
added tests for commonMinistryAdminService

min4uk committed:
added area admin tests

min4uk committed:
added update tests for ministry admin

min4uk committed:
added area admin tests

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/AreaAdminControllerTests.cs:
```diff
@@ -250,6 +252,43 @@ public async Task Update_WithErrorResponse_ReturnsStatusCodeResult()
+        // Arrange
+        var updateAreaAdminDto = new AreaAdminDto { Id = string.Empty };
+
+        areaAdminServiceMock
```
(Test logic?) Test says "_WhenUserIsNotAdmin_", but you're mocking db exception. It looks like previous test dublicate.

min4uk replied:
It is dublicate...
I'll fix it

min4uk committed:
removed excess setup
added regions
